Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp

```diff
@@ -842,6 +871,125 @@ namespace Multiplayer
+                    theirComponent.m_componentVersionHash,
+                    theirComponent.m_componentVersionHash);
+            }
+        }
+
+        // One last iteration over our components this time to check if we have a component the connected app is missing.
+        if (packet.GetTotalComponentCount() > 0)
+        {
+            for (const auto& ourComponent : GetMultiplayerComponentRegistry()->GetMultiplayerComponentVersionHashes())
+            {
+                AZ::Name ourComponentName = ourComponent.first;
+
+                bool theyHaveComponent = false;
+                for (const auto& theirComponent : theirMultiplayerComponents)
+                {
+                    if (ourComponentName == theirComponent.m_componentName)
+                    {
+                        theyHaveComponent = true;
+                        break;
+                    }
+                }
+
+                if (!theyHaveComponent)
+                {
+                    AZLOG_ERROR(
+                        "Multiplayer component mismatch! We have a component named %s which the connected application is missing!",
+                        ourComponentName.GetCStr());
+                }
+            }
+        }
+
+        // We've received all the multiplayer components for debugging.
+        // If we're the connector, send all our component information back to the acceptor.
+        if (connection->GetConnectionRole() == ConnectionRole::Connector)
+        {
+            SendAllMultiplayerComponentVersionData(connection);
+        }
+
+        // Clear out the list of components in case the other side fixes up their end (ie rebuilds using latest auto-components) and tries to reconnect
+        theirMultiplayerComponents.clear();
+        m_connectedAppsComponentVersions.erase(connection->GetConnectionId());
+
+        if (connection->GetConnectionRole() == ConnectionRole::Acceptor)
+        {
+            if (sv_versionMismatch_autoDisconnect)
+            {
+                // Disconnect from the connector; we have all the mismatch information we need now for debugging.
+                connection->Disconnect(DisconnectReason::VersionMismatch, TerminationEndpoint::Local);
+            }
+            else if (m_originalConnectPackets.contains(connection->GetConnectionId()))
+            {
+                // DANGER: We're accepting the player connection even though there's a component mismatch
+                AZLOG_WARN(
+                    "Multiplayer component mismatch was found but we're allowing the player to connect anyways. Please set sv_versionMismatch_autoDisconnect=true if this is undesired behavior!");
+                AttemptPlayerConnect(connection, m_originalConnectPackets[connection->GetConnectionId()]);
+                m_originalConnectPackets.erase(connection->GetConnectionId());
+            }
+            else
+            {
+                AZLOG_ERROR(
```

Comment:
NP: Not sure it's worth saying "please file a bug" here.

Also could this be construed as not trying to accept a specific mismatched connection with the cvar off? That use case seems rare though.

Reply:
Maybe an assert is better here? Basically if that code ever gets hit... the code is wrong and a developer needs to fix it.

Reply:
_"Also could this be construed"_
Ah I see, the logic flow was weird to read with the if/else-if/else. Updated so it's just an if/else and then another if/else inside the else, which is easier on the eyes. Lemme know!

Reply:
Replaced with assert.